crates/bevy_reflect/derive/src/enum_utility.rs

```diff
@@ -1,122 +1,263 @@
-                        }
-                    },
-                    DefaultBehavior::Default => quote! {
-                        if let #FQOption::Some(field) = #field_accessor {
-                            <#field_ty as #bevy_reflect_path::FromReflect>::from_reflect(field)
-                            #resolve_error
-                        } else {
-                            #FQDefault::default()
-                        }
-                    },
-                    DefaultBehavior::Required => quote! {
-                        <#field_ty as #bevy_reflect_path::FromReflect>::from_reflect(#field_accessor #resolve_missing)
-                        #resolve_error
-                    },
+                constructor
+            });
+
+            let constructor = quote! {
+                #variant_path {
+                    #( #field_constructors ),*
                 }
             };
-            quote! { #field_ident : #field_value }
-        });
-        variant_constructors.push(quote! {
-            #variant_constructor { #( #constructor_fields ),* }
-        });
-        variant_names.push(name);
+
+            variant_names.push(variant_name);
+            variant_constructors.push(constructor);
+        }
+
+        EnumVariantOutputData {
+            variant_names,
+            variant_constructors,
+        }
     }
 
-    EnumVariantConstructors {
-        variant_names,
-        variant_constructors,
+    fn construct_field(&self, variant_field: VariantField) -> proc_macro2::TokenStream {
+        let VariantField { field, .. } = variant_field;
+
+        // Ignored fields (fall back to default value)
+        if field.attrs.ignore.is_ignored() {
+            return match &field.attrs.default {
+                DefaultBehavior::Func(path) => quote! { #path() },
+                _ => quote! { #FQDefault::default() },
+            };
+        }
+
+        let field_accessor = (self.field_accessor)(self.this, variant_field);
+
+        let field_ident = format_ident!("__field");
+        let field_constructor = (self.field_constructor)(&field_ident, variant_field);
+
+        match &field.attrs.default {
+            DefaultBehavior::Func(path) => quote! {
+                if let #FQOption::Some(#field_ident) = #field_accessor {
+                    #field_constructor
+                } else {
+                    #path()
+                }
+            },
+            DefaultBehavior::Default => quote! {
+                if let #FQOption::Some(#field_ident) = #field_accessor {
+                    #field_constructor
+                } else {
+                    #FQDefault::default()
+                }
+            },
+            DefaultBehavior::Required => {
+                let field_unwrapper = (self.field_unwrapper)(&field_ident, variant_field);
+
+                quote! {{
+                    let #field_ident = #field_accessor;
+                    let #field_ident = #field_unwrapper;
```

Comment:
Just to be sure, it's expected to have `#field_ident` twice, because `#field_unwrapper` makes use of the first `#field_ident`?

Reply:
Yep! The `field_unwrapper` is given a `&Ident` which in this case is set to `field_ident`.

Is this just clarification for you or is there a way I could make this clearer? Maybe with a comment?

Reply:
It was mostly clarification!